File: samples/src/main/java/jvmMain/java/SafeSample.java
package jvmMain.java;

import one.mixin.bot.HttpClient;
import one.mixin.bot.api.MixinResponse;
import one.mixin.bot.extension.Base64ExtensionKt;
import one.mixin.bot.extension.ByteArrayExtensionKt;
import one.mixin.bot.extension.TimeExtensionKt;
import one.mixin.bot.safe.EdKeyPair;
import one.mixin.bot.safe.TipBody;
import one.mixin.bot.safe.TipException;
import one.mixin.bot.safe.TipKt;
import one.mixin.bot.util.ByteArrayUtilKt;
import one.mixin.bot.util.CryptoUtilKt;
import one.mixin.bot.vo.Account;
import one.mixin.bot.vo.PinRequest;
import one.mixin.bot.vo.User;

import java.io.IOException;
import java.util.Objects;

import static jvmMain.java.Sample.createUser;
import static jvmMain.java.Sample.userPin;
import static one.mixin.bot.SessionKt.encryptPin;
import static one.mixin.bot.SessionKt.encryptTipPin;

public class SafeSample {
    public static void main(String[] args) throws Exception {
        HttpClient botClient = Common.botClient;

        updateFromLegacyPin(botClient);

//         Account user = createTipPin(botClient);
    }

    private static void updateFromLegacyPin(HttpClient botClient) throws Exception {
        // create user
        EdKeyPair sessionKey = CryptoUtilKt.generateEd25519KeyPair();
        String sessionSecret = Base64ExtensionKt.base64Encode(sessionKey.getPublicKey());
        User user = createUser(botClient, sessionSecret);
        assert user != null;

        HttpClient userClient = new HttpClient.Builder().useCNServer().enableDebug().configSafeUser(
                user.getUserId(),
                user.getSessionId(),
                sessionKey.getPrivateKey(),
                null,
                null
        ).build();

        // decrypt pin token
        byte[] userPrivateKey = sessionKey.getPrivateKey();
        byte[] userAesKey = CryptoUtilKt.decryptPinToken(Base64ExtensionKt.base64Decode(user.getPinToken()), userPrivateKey);

        // create user pin
        MixinResponse<User> response = userClient.getUserService().createPinCall(
                new PinRequest(encryptPin(userAesKey, userPin), null, null)
        ).execute().body();
        assert response != null;
        if (response.isSuccess()) {
            System.out.println("Create pin success " + response.getData());
        } else {
            throw new Exception("Create pin failure " + response.getError());
        }
        // verify usr pin
        response = userClient.getUserService().pinVerifyCall(
                new PinRequest(encryptPin(userAesKey, userPin), null, null)
        ).execute().body();
        assert response != null;
        if (response.isSuccess()) {
            System.out.println("Verify pin success");
        } else {
            throw new Exception("Verify pin failure " + response.getError());
        }

        // update tip pin
        byte[] tipSeed = CryptoUtilKt.generateRandomBytes(32);
        EdKeyPair keyPair = CryptoUtilKt.newKeyPairFromSeed(tipSeed);
        TipKt.updateTipPin(userClient, ByteArrayExtensionKt.toHex(keyPair.getPublicKey()), Base64ExtensionKt.base64UrlEncode(userPrivateKey), user.getPinToken(), userPin);
        // verify tip pin
        long timestamp = TimeExtensionKt.nowInUtcNano();
        response = userClient.getUserService().pinVerifyCall(
                new PinRequest(encryptTipPin(userAesKey, TipBody.forVerify(timestamp), keyPair.getPrivateKey()), null, timestamp)
        ).execute().body();
        assert response != null;
        if (response.isSuccess()) {
            System.out.println("Verify tip pin success");
        } else {
            throw new Exception("Verify tip pin failure " + response.getError());
        }

        // register safe
        TipKt.registerSafe(userClient, user.getUserId(), ByteArrayExtensionKt.toHex(keyPair.getPrivateKey()), ByteArrayExtensionKt.toHex(keyPair.getPrivateKey()), Base64ExtensionKt.base64UrlEncode(userPrivateKey), user.getPinToken());
    }

    private static Account createTipPin(HttpClient botClient) throws IOException, TipException {
        // create user
        EdKeyPair sessionKey = CryptoUtilKt.generateEd25519KeyPair();
        String sessionSecret = Base64ExtensionKt.base64Encode(sessionKey.getPublicKey());
        User user = createUser(botClient, sessionSecret);
        assert user != null;

        HttpClient userClient = new HttpClient.Builder().useCNServer().enableDebug().configSafeUser(
                user.getUserId(),
                user.getSessionId(),
                sessionKey.getPrivateKey(),
                null,
                null
        ).build();

        // decrypt pin token
        byte[] userPrivateKey = sessionKey.getPrivateKey();
        byte[] userAesKey = CryptoUtilKt.decryptPinToken(Base64ExtensionKt.base64Decode(user.getPinToken()), userPrivateKey);

        // create user tip pin
        byte[] tipSeed = CryptoUtilKt.generateRandomBytes(32);
        EdKeyPair keyPair = CryptoUtilKt.newKeyPairFromSeed(tipSeed);
        MixinResponse<User> response = userClient.getUserService().createPinCall(
                new PinRequest(
                        encryptPin(userAesKey, Bytes.concat(keyPair.getPublicKey(), ByteArrayUtilKt.toBeByteArray(1L))),
                        null, null
                )).execute().body();
        assert response != null;
        if (response.isSuccess()) {
            System.out.println("Create tip pin success " + Objects.requireNonNull(response.getData()));
        } else {
            System.out.println("Create tip pin failure " + response.getError());
        }

        // register safe
        return TipKt.registerSafe(userClient, user.getUserId(), ByteArrayExtensionKt.toHex(keyPair.getPrivateKey()), ByteArrayExtensionKt.toHex(keyPair.getPrivateKey()), Base64ExtensionKt.base64UrlEncode(userPrivateKey), user.getPinToken());
    }
}

File: samples/src/main/java/jvmMain/kotlin/SafeSample.kt
package jvmMain.kotlin

import kotlinx.coroutines.runBlocking
import one.mixin.bot.HttpClient
import one.mixin.bot.encryptPin
import one.mixin.bot.encryptTipPin
import one.mixin.bot.extension.base64Decode
import one.mixin.bot.extension.base64Encode
import one.mixin.bot.extension.base64UrlEncode
import one.mixin.bot.extension.nowInUtcNano
import one.mixin.bot.extension.toHex
import one.mixin.bot.safe.TipBody
import one.mixin.bot.safe.registerSafe
import one.mixin.bot.safe.updateTipPin
import one.mixin.bot.util.decryptPinToken
import one.mixin.bot.util.generateEd25519KeyPair
import one.mixin.bot.util.generateRandomBytes
import one.mixin.bot.util.newKeyPairFromSeed
import one.mixin.bot.util.toBeByteArray
import one.mixin.bot.vo.Account
import one.mixin.bot.vo.PinRequest

fun main() = runBlocking {
   updateFromLegacyPin(botClient)

   // val user = createTipPin(botClient) ?: return@runBlocking
}

private suspend fun updateFromLegacyPin(botClient: HttpClient) { // create user
   val sessionKey = generateEd25519KeyPair()
   val sessionSecret = sessionKey.publicKey.base64Encode()
   val user = createUser(botClient, sessionSecret) ?: return

   val userClient = HttpClient.Builder().useCNServer().configSafeUser(
       user.userId,
       user.sessionId,
       sessionKey.privateKey,
   ).enableDebug().build()

   // decrypt pin token
   val userPrivateKey = sessionKey.privateKey
   val userAesKey = decryptPinToken(user.pinToken.base64Decode(), userPrivateKey)

   // create user pin
   var response = userClient.userService.createPin(PinRequest(encryptPin(userAesKey, DEFAULT_PIN.toByteArray())))
   if (response.isSuccess()) {
       println("Create pin success ${response.data}")
   } else {
       throw Exception("Create pin failure ${response.error}")
   } // verify usr pin
   response = userClient.userService.pinVerify(PinRequest(encryptPin(userAesKey, DEFAULT_PIN.toByteArray())))
   if (response.isSuccess()) {
       println("Verify pin success")
   } else {
       throw Exception("Verify pin failure ${response.error}")
   }

   // update tip pin
   val tipSeed = generateRandomBytes(32)
   val keyPair = newKeyPairFromSeed(tipSeed)
   updateTipPin(
       userClient,
       keyPair.publicKey.toHex(),
       userPrivateKey.base64UrlEncode(),
       user.pinToken,
       DEFAULT_PIN
   ) // verify tip pin
   val timestamp = nowInUtcNano()
   response = userClient.userService.pinVerify(
       PinRequest(
           encryptTipPin(
               userAesKey,
               TipBody.forVerify(timestamp),
               keyPair.privateKey
           ), timestamp = timestamp
       )
   )
   if (response.isSuccess()) {
       println("Verify tip pin success")
   } else {
       throw Exception("Verify tip pin failure ${response.error}")
   }

   // register safe
   registerSafe(
       userClient,
       user.userId,
       keyPair.privateKey.toHex(),
       keyPair.privateKey.toHex(),
       userPrivateKey.base64UrlEncode(),
       user.pinToken,
   )
}

private suspend fun createTipPin(botClient: HttpClient): Account? { // create user
   val sessionKey = generateEd25519KeyPair()
   val sessionSecret = sessionKey.publicKey.base64Encode()
   val user = createUser(botClient, sessionSecret) ?: return null

   val userClient = HttpClient.Builder().useCNServer().configSafeUser(
       user.userId,
       user.sessionId,
       sessionKey.privateKey,
   ).enableDebug().build()

   // decrypt pin token
   val userPrivateKey = sessionKey.privateKey
   val userAesKey = decryptPinToken(user.pinToken.base64Decode(), userPrivateKey)

   // create user tip pin
   val tipSeed = generateRandomBytes(32)
   val keyPair = newKeyPairFromSeed(tipSeed)
   val response = userClient.userService.createPin(
       PinRequest(encryptPin(userAesKey, keyPair.publicKey + 1L.toBeByteArray())),
   )
   if (response.isSuccess()) {
       println("Create tip pin success ${response.data?.userId}")
   } else {
       println("Create tip pin failure ${response.error}")
   }

   // register safe
   return registerSafe(
       userClient,
       user.userId,
       keyPair.privateKey.toHex(),
       keyPair.privateKey.toHex(),
       userPrivateKey.base64UrlEncode(),
       user.pinToken,
   )
}